core_impl/src/main/java/com/google/instrumentation/stats/StatsSerializer.java

```diff
@@ -31,52 +29,81 @@
+  //         <tag_val_len> == varint encoded integer
+  //         <tag_val> == tag_val_len bytes comprising UTF-8 string
+  //     <tag_encoding> (tag_field_id == 1) ==
+  //       <tag_key_len><tag_key><int_tag_val>
+  //         <tag_key_len> == varint encoded integer
+  //         <tag_key> == tag_key_len bytes comprising tag key name
+  //         <int_tag_value> == 8 bytes, little-endian integer
+  //     <tag_encoding> (tag_field_id == 2 || tag_field_id == 3) ==
+  //       <tag_key_len><tag_key>
+  //         <tag_key_len> == varint encoded integer
+  //         <tag_key> == tag_key_len bytes comprising tag key name
   //    TODO(songya): Currently we only support encoding on string type.
+  private static final int VERSION_ID = 0;
   private static final int VALUE_TYPE_STRING = 0;
   private static final int VALUE_TYPE_INTEGER = 1;
-  private static final int VALUE_TYPE_BOOLEAN = 2;
+  private static final int VALUE_TYPE_TRUE = 2;
+  private static final int VALUE_TYPE_FALSE = 3;
 
-  // Serializes a StatsContext by transforming it into a StatsContextProto. The
-  // encoded tags are of the form:
-  //   [tag_metadata key_len key_bytes value_len value_bytes]*
+
+  // Serializes a StatsContext to the on-the-wire format.
+  // Encoded tags are of the form: <version_id><encoded_tags>
   static void serialize(StatsContextImpl context, OutputStream output) throws IOException {
     ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    VarInt.putVarInt(VERSION_ID, byteArrayOutputStream);
 
     // TODO(songya): add support for value types integer and boolean
     Set<Entry<String, String>> tags = context.tags.entrySet();
     for (Entry<String, String> tag : tags) {
-      encodeTagWithType(VALUE_TYPE_STRING, tag.getKey(), tag.getValue(), byteArrayOutputStream);
+      encodeStringTagWithType(
+          VALUE_TYPE_STRING, tag.getKey(), tag.getValue(), byteArrayOutputStream);
     }
-    ByteString encodedTags = ByteString.copyFrom(byteArrayOutputStream.toByteArray());
-    StatsContextProto.StatsContext.newBuilder().setTags(encodedTags).build().writeTo(output);
+    byteArrayOutputStream.writeTo(output);
   }
 
-  // Deserializes based on an serialized StatsContextProto. The encoded tags are of the form:
-  //   [tag_metadata key_len key_bytes value_len value_bytes]*
+  // Deserializes input to StatsContext based on the binary format standard.
+  // The encoded tags are of the form: <version_id><encoded_tags>
   static StatsContextImpl deserialize(InputStream input) throws IOException {
     try {
-      StatsContextProto.StatsContext context = StatsContextProto.StatsContext.parseFrom(input);
-      ByteBuffer buffer = context.getTags().asReadOnlyByteBuffer();
+      byte[] bytes = ByteStreams.toByteArray(input);
       HashMap<String, String> tags = new HashMap<String, String>();
+      if (bytes.length == 0) {
+        // Return a default StatsContext on empty input.
```

Comment:
for empty byte array I would return a ParseException. Also I think it is better to use ParseException instead of IOException to be consistent with the trace.